core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -2032,4 +2001,207 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  private[kafka] def getOrCreatePartition(tp: TopicPartition,
+                                          delta: TopicsDelta,
+                                          topicId: Uuid): Option[(Partition, Boolean)] = {
+    getPartition(tp) match {
+      case HostedPartition.Offline =>
+        stateChangeLogger.warn(s"Unable to bring up new local leader ${tp} " +
+          s"with topic id ${topicId} because it resides in an offline log " +
+          "directory.")
+        None
+
+      case HostedPartition.Online(partition) => Some(partition, false)
```

Comment:
Should we validate the topicId matches?

Reply:
It's a bit messy because actually the topic ID is not stored in the Partition object, but in the Log object, which may not actually exist (although we'd want to transition out of online mode eventually if that were the case, of course.) I will add a check that if the id is defined it's correct, though.